From 2f07cdd8de077db287e0579adb5fb1f288fa3afe Mon Sep 17 00:00:00 2001
From: Rahul Sandhu <nvraxn@gmail.com>
Date: Mon, 3 Nov 2025 15:31:27 +0000
Subject: [PATCH 08/17] wayland: allow violating W^X for compositors behind a
 boolean

This is probably to do with llvmpipe. Further analysis to come later.

Signed-off-by: Rahul Sandhu <nvraxn@gmail.com>
---
 policy/global_tunables            | 8 ++++++++
 policy/modules/session/wayland.te | 5 +++++
 2 files changed, 13 insertions(+)

diff --git a/policy/global_tunables b/policy/global_tunables
index 1dd585220..64342ae32 100644
--- a/policy/global_tunables
+++ b/policy/global_tunables
@@ -137,3 +137,11 @@ gen_tunable(user_udp_server,false)
 ## </p>
 ## </desc>
 gen_tunable(user_execmod_tmpfs,false)
+
+## <desc>
+## <p>
+## Allow Wayland compositors to execute writeable memory
+## regions.  This is likely needed for llvmpipe to work.
+## </p>
+## </desc>
+gen_tunable(wayland_compositor_execmem,false)
diff --git a/policy/modules/session/wayland.te b/policy/modules/session/wayland.te
index 8b0cc2d57..395a911d8 100644
--- a/policy/modules/session/wayland.te
+++ b/policy/modules/session/wayland.te
@@ -90,6 +90,11 @@ xserver_use_user_fonts(wayland_compositor)
 xserver_read_xkb_libs(wayland_compositor)
 xserver_rw_mesa_shader_cache(wayland_compositor)
 
+tunable_policy(`wayland_compositor_execmem',`
+	# Most likely needed for llvmpipe.
+	allow wayland_compositor self:process execmem;
+')
+
 optional_policy(`
 	# For seats
 	systemd_dbus_chat_logind(wayland_compositor)
-- 
2.47.3

